File: Linear_Data_Structures/Stack/stack_linked_list.py
"""
Stack implementation using a linked list (node-based).
Offers fine-grained memory control and no amortized resizing overhead.
Time Complexity: Push O(1), Pop O(1), Peek O(1)
"""


class Node:
    """Node for the linked list."""
    
    def __init__(self, data):
        """Initialize a node with data and next pointer."""
        self.data = data
        self.next = None


class Stack:
    """Stack implementation using a linked list."""
    
    def __init__(self):
        """Initialize an empty stack."""
        self.top = None
        self.size_count = 0

    def push(self, data):
        """Insert an element at the top."""
        new_node = Node(data)
        new_node.next = self.top
        self.top = new_node
        self.size_count += 1

    def pop(self):
        """Remove and return the top element."""
        if self.top is None:
            return None
        popped = self.top.data
        self.top = self.top.next
        self.size_count -= 1
        return popped

    def peek(self):
        """View the top element without removing."""
        if self.top is None:
            return None
        return self.top.data

    def is_empty(self):
        """Check if the stack is empty."""
        return self.top is None

    def size(self):
        """Get the number of elements."""
        return self.size_count

    def __str__(self):
        """String representation of the stack."""
        elements = []
        current = self.top
        while current:
            elements.append(str(current.data))
            current = current.next
        return f"Stack([{' -> '.join(elements)}])"


if __name__ == "__main__":
    # Demo
    stack = Stack()
    print("Empty stack:", stack)
    
    print("\nPushing 10, 20, 30:")
    stack.push(10)
    stack.push(20)
    stack.push(30)
    print(stack)
    
    print("\nPeek:", stack.peek())
    print("Size:", stack.size())
    
    print("\nPopping:")
    while not stack.is_empty():
        print(f"  Popped: {stack.pop()}")
    
    print("Empty again:", stack.is_empty())

File: Linear_Data_Structures/Stack/stack_linked_list.c
#include <stdio.h>
#include <stdlib.h>

/*
 * Stack implementation using linked list (node-based).
 * Offers fine-grained memory control and no resizing overhead.
 * Time Complexity: Push O(1), Pop O(1), Peek O(1)
 */

typedef struct Node {
    int data;
    struct Node *next;
} Node;

typedef struct {
    Node *top;
    int size_count;
} Stack;

Stack* create_stack() {
    Stack *s = (Stack *)malloc(sizeof(Stack));
    s->top = NULL;
    s->size_count = 0;
    return s;
}

Node* create_node(int data) {
    Node *n = (Node *)malloc(sizeof(Node));
    n->data = data;
    n->next = NULL;
    return n;
}

void push(Stack *s, int data) {
    Node *new_node = create_node(data);
    new_node->next = s->top;
    s->top = new_node;
    s->size_count++;
}

int pop(Stack *s) {
    if (s->top == NULL) {
        printf("Stack underflow!\n");
        return -1;
    }
    int popped = s->top->data;
    Node *temp = s->top;
    s->top = s->top->next;
    free(temp);
    s->size_count--;
    return popped;
}

int peek(Stack *s) {
    if (s->top == NULL) {
        return -1;
    }
    return s->top->data;
}

int is_empty(Stack *s) {
    return s->top == NULL;
}

int size(Stack *s) {
    return s->size_count;
}

void print_stack(Stack *s) {
    printf("[");
    Node *current = s->top;
    int first = 1;
    while (current != NULL) {
        if (!first) printf(" -> ");
        printf("%d", current->data);
        current = current->next;
        first = 0;
    }
    printf("]\n");
}

void free_stack(Stack *s) {
    while (!is_empty(s)) {
        pop(s);
    }
    free(s);
}

int main() {
    Stack *stack = create_stack();
    printf("Empty stack: ");
    print_stack(stack);
    
    printf("\nPushing 10, 20, 30:\n");
    push(stack, 10);
    push(stack, 20);
    push(stack, 30);
    printf("Stack: ");
    print_stack(stack);
    
    printf("Peek: %d\n", peek(stack));
    printf("Size: %d\n", size(stack));
    
    printf("\nPopping:\n");
    while (!is_empty(stack)) {
        printf("  Popped: %d\n", pop(stack));
    }
    
    printf("Empty again: %s\n", is_empty(stack) ? "true" : "false");
    
    free_stack(stack);
    return 0;
}
